azurelinuxagent/ga/policy/regorus.py

```diff
@@ -0,0 +1,169 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.4+ and Openssl 1.0+
+#
+
+import json
+import os
+import tempfile
+import azurelinuxagent.common.utils.shellutil as shellutil
+from azurelinuxagent.common.utils.shellutil import CommandError
+from azurelinuxagent.common.exception import PolicyError
+
+
+
+def get_regorus_path():
+    """
+    Returns path to Regorus executable. The executable is not yet officially released as part of the agent package.
+    After release, executable will be placed at /var/lib/waagent/WALinuxAgent-x.x.x.x/bin/regorus.
+    Currently, the executable is copied into the agent directory for unit testing, and this method is mocked.
+    """
+    regorus_exe = os.path.join(os.getcwd(), "bin", "regorus")
+    return regorus_exe
+
+
+class Engine:
+    """
+    This class implements the basic operations for the Regorus policy engine via subprocess.
+    Any errors thrown in this class should be caught and handled by PolicyEngine.
+    """
+
+    def __init__(self):
+        self._engine = None
+        self._rule_file = None
+        self._policy_file = None
+        self._input_json = None
+
+    def add_policy(self, rule_file):
+        """Rule_file is expected to point to a valid Regorus file."""
+        if not os.path.exists(rule_file):
+            raise IOError("Policy rule file path {0} does not exist".format(rule_file))
+        if not rule_file.endswith('.rego'):
+            raise TypeError("Policy rule file path {0} is not a valid .rego file.".format(rule_file))
+        self._rule_file = rule_file
+        # TODO: Currently, eval_query will raise error if policy file is invalid.
+        #  Consider a dummy "regorus eval" or "regorus parse" call here to validate policy rule file beforehand.
+
+    def set_input(self, input_json):
+        """
+        This sets the input to the policy engine query, for example, a list of extensions we want to download.
+        Input_json should be a JSON object.
+        Expected format:
+        {
+            "extensions": {
+                "<extension_name_1>": {
+                    "signingInfo": {
+                        "extensionSigned": <true, false>
+                    }
+                },
+                "<extension_name_2>": {
+                    "signingInfo": {
+                        "extensionSigned": <true, false>
+                    }
+                }, ...
+        }
+        """
+        if isinstance(input_json, dict):
+            self._input_json = input_json
+        elif isinstance(input_json, str):
+            self._input_json = json.loads(input_json)
+        else:
+            raise TypeError("Input to policy engine is not a valid JSON object.")
+        # TODO: Add JSON schema and validate input against it, return detailed error message
```

Comment:
Do we need to do our own validation here? wouldn't the regorus engine do its own validation?

Reply:
same comment for line 109 below

Reply:
Regorus engine does do its own validation, but the error messages/codes that are popped up are not very clear to the user. The intention behind pre-validating the input is to clearly identify any issues and provide detailed error messages. 

But I am removing these validations now. We can request the Regorus team to improve clarity of error messages and return codes.

Reply:
Done